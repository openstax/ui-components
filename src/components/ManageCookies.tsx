import React from 'react';
import { ButtonLink } from "./Button";
import { createGlobalStyle } from "styled-components";

const GlobalStyle = createGlobalStyle`
  .cky-btn-revisit { display: none; }
`;

type ManageCookiesLinkProps = React.ButtonHTMLAttributes<HTMLButtonElement> & {
  wrapper?: (button: React.ReactElement) => React.ReactElement;
};

// documentation for this at https://www.cookieyes.com/documentation/change-cookie-consent-using-cookieyes/
export const ManageCookiesLink = ({children, className, wrapper, ...props}: ManageCookiesLinkProps) => {
  const inBrowser = typeof window === 'object';
  const [cookieYesLoaded, setCookieYesLoaded] = React.useState(false);

  React.useEffect(() => {
    if (inBrowser && !cookieYesLoaded) {
      const handleCkyLoaded = () => setCookieYesLoaded(true);
      if ('getCkyConsent' in window) {
        handleCkyLoaded();
      } else {
        document.addEventListener('cookieyes_banner_load', handleCkyLoaded);
        return () => document.removeEventListener('cookieyes_banner_load', handleCkyLoaded);
      }
    }
    return;
  }, [inBrowser]);

  if (!inBrowser) { return <><GlobalStyle /></>; }

  const button = <ButtonLink
    className={`cky-banner-element${className ? ` ${className}` : ''}`}
    {...props}
  >{children || 'Manage Cookies'}</ButtonLink>;

  return <>
    <GlobalStyle />
    {cookieYesLoaded
       ? typeof wrapper === 'function'
         ? wrapper(button)
         : button
       : null
    }
  </>;
};
