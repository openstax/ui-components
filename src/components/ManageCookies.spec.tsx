/* eslint-disable @typescript-eslint/no-explicit-any */
import renderer from 'react-test-renderer';
import { ManageCookiesLink } from "./ManageCookies";

describe('ManageCookies', () => {
  describe('without osano', () => {

    it('renders nothing', () => {
      const tree = renderer.create(
        <ManageCookiesLink />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders nothing, even a wrapper', () => {
      const tree = renderer.create(
        <ManageCookiesLink>{button => <div>{button}</div>}</ManageCookiesLink>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('with osano', () => {
    let showDrawerSpy: jest.SpyInstance;

    beforeAll(() => {
      showDrawerSpy = jest.fn();
      (window as any).Osano = {cm: {mode: 'production',
        showDrawer: showDrawerSpy}
      };
    });

    afterAll(() => {
      delete (window as any).Osano;
    });

    beforeEach(() => {
      jest.clearAllMocks();
    });

    it('renders button', () => {
      const tree = renderer.create(
        <ManageCookiesLink />
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('renders button in wrapper', () => {
      const tree = renderer.create(
        <ManageCookiesLink>{button => <div>{button}</div>}</ManageCookiesLink>
      ).toJSON();
      expect(tree).toMatchSnapshot();
    });

    it('calls osano showDrawer', () => {
      const component = renderer.create(
        <ManageCookiesLink />
      );

      expect(showDrawerSpy).not.toHaveBeenCalled();
      component.root.findByType('button').props.onClick();
      expect(showDrawerSpy).toHaveBeenCalled();
    });

    it('calls additional callback', () => {
      const onClick = jest.fn();
      const component = renderer.create(
        <ManageCookiesLink onClick={onClick} />
      );

      expect(showDrawerSpy).not.toHaveBeenCalled();
      component.root.findByType('button').props.onClick();
      expect(showDrawerSpy).toHaveBeenCalled();
      expect(onClick).toHaveBeenCalled();
    });
  });
});
