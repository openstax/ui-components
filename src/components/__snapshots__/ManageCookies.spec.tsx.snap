// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageCookies when CookieYes loads renders button 1`] = `
<button
  className="sc-gsnTZi sc-dkzDqf czjxlp jmylct cky-banner-element"
>
  Manage Cookies
</button>
`;

exports[`ManageCookies when CookieYes loads renders button in wrapper 1`] = `
<div>
  <button
    className="sc-gsnTZi sc-dkzDqf czjxlp jmylct cky-banner-element"
  >
    Manage Cookies
  </button>
</div>
`;

exports[`ManageCookies when CookieYes loads renders button with content and correct class 1`] = `
<button
  className="sc-gsnTZi sc-dkzDqf czjxlp jmylct cky-banner-element"
>
  some content
</button>
`;

exports[`ManageCookies with CookieYes already loaded renders button 1`] = `
<button
  className="sc-gsnTZi sc-dkzDqf czjxlp jmylct cky-banner-element"
>
  Manage Cookies
</button>
`;

exports[`ManageCookies with CookieYes already loaded renders button in wrapper 1`] = `
<div>
  <button
    className="sc-gsnTZi sc-dkzDqf czjxlp jmylct cky-banner-element"
  >
    Manage Cookies
  </button>
</div>
`;

exports[`ManageCookies with CookieYes already loaded renders button with content and correct class 1`] = `
<button
  className="sc-gsnTZi sc-dkzDqf czjxlp jmylct cky-banner-element"
>
  some content
</button>
`;

exports[`ManageCookies without cookieYes renders nothing 1`] = `null`;

exports[`ManageCookies without cookieYes renders nothing, even a wrapper 1`] = `null`;
